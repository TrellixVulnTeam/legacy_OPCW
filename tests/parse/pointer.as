
# Pointers are declared using the C-style "*" symbol but as a prefix to
# the type rather than a suffix.
let x: *int;
let x: ***int;

# Pointers and their values are both immutable by default.
let mut x: *int;        # mutable pointer to an immutable value
let x: *mut int;        # immutable pointer to a mutable value
let mut x: *mut int;    # mutable pointer to a mutable value
let x: *mut *int;
let x: *mut *mut int;
let x: *mut *mut *int;
let x: *mut *mut *mut int;

# An address of an existing value can be taken with the "&" symbol prefix; this
# can create a pointer (much like in "C").
# NOTE: An immutable pointer is inferred by default; use "&mut " to take a
#       "mutable" address.
let y = 73;
let x = &y;
let a = 73;
let x = &mut a;
let x: *mut int = &a;  # semantic error because &a is '*int' not '*mut int'

# The value of a pointer can be loaded by the unary "*" operator (
# much like in "C").
let x: int;
let y: *int = &x;
let z: int = *y;
let w: int = *&y;
