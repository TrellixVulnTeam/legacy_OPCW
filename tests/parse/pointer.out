ModuleDecl <?> _
  Nodes <?> 
    LocalSlotDecl <?> x
      PointerType <?>
        Ident <?> int
    LocalSlotDecl <?> x
      PointerType <?>
        PointerType <?>
          PointerType <?>
            Ident <?> int
    LocalSlotDecl <?> mut x
      PointerType <?>
        Ident <?> int
    LocalSlotDecl <?> x
      PointerType <?> mut
        Ident <?> int
    LocalSlotDecl <?> mut x
      PointerType <?> mut
        Ident <?> int
    LocalSlotDecl <?> x
      PointerType <?> mut
        PointerType <?>
          Ident <?> int
    LocalSlotDecl <?> x
      PointerType <?> mut
        PointerType <?> mut
          Ident <?> int
    LocalSlotDecl <?> x
      PointerType <?> mut
        PointerType <?> mut
          PointerType <?>
            Ident <?> int
    LocalSlotDecl <?> x
      PointerType <?> mut
        PointerType <?> mut
          PointerType <?> mut
            Ident <?> int
    LocalSlotDecl <?> y
      IntegerExpr <?> 73 (10)
    LocalSlotDecl <?> x
      AddressOfExpr <?>
        Ident <?> y
    LocalSlotDecl <?> a
      IntegerExpr <?> 73 (10)
    LocalSlotDecl <?> x
      AddressOfExpr <?> mut
        Ident <?> a
    LocalSlotDecl <?> x
      PointerType <?> mut
        Ident <?> int
      AddressOfExpr <?>
        Ident <?> a
    LocalSlotDecl <?> x
      Ident <?> int
    LocalSlotDecl <?> y
      PointerType <?>
        Ident <?> int
      AddressOfExpr <?>
        Ident <?> x
    LocalSlotDecl <?> z
      Ident <?> int
      DerefExpr <?>
        Ident <?> y
    LocalSlotDecl <?> w
      Ident <?> int
      DerefExpr <?>
        AddressOfExpr <?>
          Ident <?> y
